dbms/src/Server/Server.cpp

```diff
@@ -1250,9 +1247,51 @@ try
             LOG_ERROR(log, "CPUAffinityManager::bindThreadCPUAffinity throws exception.");
         }
 
+        // Ready to provide service
+        tmt_context.setStatusRunning();
+
         LOG_INFO(log, "Start to wait for terminal signal");
         waitForTerminationRequest();
 
+        LOG_INFO(log, "Set store context status Stopping");
+        tmt_context.setStatusStopping();
+
+        // Wait for all mpp tasks to finish
+        auto monitor = tmt_context.getMPPTaskManager()->getMPPTaskMonitor();
+        auto start = std::chrono::steady_clock::now();
+        // The maximum seconds TiFlash will wait for all current MPP tasks to finish before shutting down
+        static constexpr const char * GRACEFUL_WIAT_BEFORE_SHUTDOWN = "flash.graceful_wait_before_shutdown";
+        auto graceful_wait_before_shutdown
+            = global_context->getUsersConfig()->getUInt64(GRACEFUL_WIAT_BEFORE_SHUTDOWN, 60);
+        auto max_wait_time = start + std::chrono::seconds(graceful_wait_before_shutdown);
+        // Sleep before checking to reduce the chance of missing MPP tasks that are still in the process of being dispatched
+        std::this_thread::sleep_for(std::chrono::milliseconds(200));
+        while (true)
+        {
```

Comment:
maybe we can use 
```
do {
  std::this_thread::sleep_for(std::chrono::milliseconds(200));
  // check whether mpp task are all finished
} while (...)
```
instead of writing `std::this_thread::sleep_for(std::chrono::milliseconds(200));` twice

Reply:
Done.